<a href="https://colab.research.google.com/github/TheMaze45/Pandas/blob/main/Recap_Data_Exploration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Recap on Pandas DataFrames

DataFrames in Pandas are mutable two-dimensional structures of data with labeled axes where:



*   each row represents a different observation
*   each column represents a different variable



Before we can create a DataFrame, we have to import the Pandas module into 
our coding enviroment.

In [2]:
# We can do that with
import pandas as pd 
# you can name pandas anything you want, but importing pandas as pd is the standard !

# 1. Import a CSV-file (comma seperated values) into the DataFrame


When working with data, most of time you don't create it by yourself, your main job as a Data-Analyst/Scientist is to get data and work with it.

We call this process the importing(or "reading") of data, sometimes from a CSV-file or from a Database.

With Pandas it is easy to import data and convert it into a DataFrame.

Here we take for example a CSV-file that we worked with before.

##Important:

When importing a file from your Google Drive, make sure that the access permission for the file is set to "anyone with the link can read/view the file",otherwise you will have trouble importing the data.

In [3]:
# url is just the link path to your file
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing" # orderlines.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
# This here creates the dataframe
df = pd.read_csv(path)

# Take a quick glance at the dataframe to see if the import worked.
# Don't forget to run the import pandas as pd first !
df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

As we can see, the table looks good, we have valid data for every column and row.

#2. DataFrame dimensions

If you don't want or can't count the number of rows and columns by hand on your DataFrame, don't worry, there is a nice little helper



```
df.shape
```
With this attribute we can a quick overview of the size of our table.


In [4]:
df.shape

(293983, 7)

It returns a [tuple](https://www.w3schools.com/python/python_tuples.asp) (data type with multiple items stored in a single var).

The first element of our tuple is the number of rows in our DataFrame, which is 293983 rows, the second element is the amount of columns, for our example it is 7.

If we want to present this kind of information in a nice way, we could extract the values and put the information into a print statement.

In [5]:
nrows = df.shape[0]
ncols = df.shape[1]
print(f'The number of rows in our dataframe is{nrows} and we have a total number of {ncols} columns.')

The number of rows in our dataframe is293983 and we have a total number of 7 columns.


The [DataFrame.size](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.size.html) returns the total number of values that a DataFrame has.

In our case it would be the number of rows (293983) * by the number of columns (7). So a total of 2057881 values.

In [6]:
df.size

2057881

We can verify this with a boolean statement.


```
df.shape[0] * df.shape[1] == df.size
```
If the numbers match, it should return True.


In [7]:
df.shape[0] * df.shape[1] == df.size

True

With the [DataFrame.ndim](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ndim.html) attribute we can calculate the number of dimensions.
A DataFrame always has two dimensions, because it consists of rows & columns.

A series on the other hand would only have one dimension.

In [18]:
df.ndim

2

In [16]:
# Create example series to show the difference
test_series = pd.Series({"Country":"Germany","Capital":"Berlin","Residents":83200000})
test_series

Country       Germany
Capital        Berlin
Residents    84000000
dtype: object

In [17]:
test_series.ndim

1

#3. DataFrame exploration

As already said above [DataFrame.head()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html) and [DataFrame.tail()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html) are quite nice and quick ways to get a first glance at the DataFrame.

By default the first 5 / last 5 rows will be shown.

You can ofc, change that, if you give the function an argument.


In [19]:
# Without arguments
df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [20]:
# Without arguments
df.tail()

id  id_order  product_id  product_quantity      sku unit_price  \
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36

In [21]:
# Display the first 10 entries
df.head(10)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   
9  1119120    295347           0                 1  APP0700      72.19   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38  
5  2017-01-01 01:14:27  
6  2017-01-01 01:17:21  
7  2017-01-01 01:46:16  
8  2017-01-01 01:50:34  
9  2017-01-01 01:54:11

In [22]:
# Display the last 2 entries
df.tail(2)

id  id_order  product_id  product_quantity      sku unit_price  \
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36

### General info about our DataFrame

With the methods 

*   [DataFrame.info()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html)
*   [DataFrame.describe()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html)
*   [DataFrame.nunique()](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.nunique.html)

We can have a general overview of what is inside our DataFrame





```
df.info() 
```
This method tells us:

*   How the data is stored
*   Whether there are any missing values
*   How many rows and columns exist in our DataFrame


In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB




```
df.describe()
```
This method gives us an overview of the [descriptive statistics](https://www.scribbr.com/statistics/descriptive-statistics/) for the ***numerical columns of our DataFrame***.

In [24]:
df.describe()

id       id_order  product_id  product_quantity
count  2.939830e+05  293983.000000    293983.0     293983.000000
mean   1.397918e+06  419999.116544         0.0          1.121126
std    1.530096e+05   66344.486479         0.0          3.396569
min    1.119109e+06  241319.000000         0.0          1.000000
25%    1.262542e+06  362258.500000         0.0          1.000000
50%    1.406940e+06  425956.000000         0.0          1.000000
75%    1.531322e+06  478657.000000         0.0          1.000000
max    1.650203e+06  527401.000000         0.0        999.000000



```
df.nunique()
```

This method can show us the unique values counted in our DataFrame.


In [5]:
df.nunique()

id                  293983
id_order            204855
product_id               1
product_quantity        67
sku                   7951
unit_price           11329
date                251631
dtype: int64

In [7]:
# For a single column
df["id_order"].nunique()

204855

If you wish to store the unique values from a column in a "numpy array" you can do this with the [.unique()](https://pandas.pydata.org/docs/reference/api/pandas.Series.unique.html) method.
This method returns the unique values from a column.

In [15]:
df["sku"].unique()[:5] 
# The [:5] means we only want the FIRST 5 entries in that column
# [5:] would mean the last 5 columns

array(['OTT0133', 'LGE0043', 'PAR0071', 'WDT0315', 'JBL0104'],
      dtype=object)



```
df.isna()
```

The [DataFrame.isna()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isna.html) method returns a boolean for each value in our DataFrame.


*   `True` if that value is "missing" meaning `NaN`
*   `False` if the value is not missing

In [19]:
df.isna()

id  id_order  product_id  product_quantity    sku  unit_price  \
0       False     False       False             False  False       False   
1       False     False       False             False  False       False   
2       False     False       False             False  False       False   
3       False     False       False             False  False       False   
4       False     False       False             False  False       False   
...       ...       ...         ...               ...    ...         ...   
293978  False     False       False             False  False       False   
293979  False     False       False             False  False       False   
293980  False     False       False             False  False       False   
293981  False     False       False             False  False       False   
293982  False     False       False             False  False       False   

         date  
0       False  
1       False  
2       False  
3       False  
4       False  
...       ...  
293978  False  
293979  False  
293980  False  
293981  False  
293982  False  

[293983 rows x 7 columns]

We can combine this method with [DataFrame.sum()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html) to add up all the missing booleans for each column and then count how many rows are missing values.

Remember:



*   `True` is interpreted as `1` meaning missing data
*   `False` is interpreted as `0` meaning we have data



In [20]:
df.isna().sum()

id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64

When we get back only 0, we should be happy, in all our columns all only rows that have a complete set of data !

[DataFrame.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) also returns a boolean output, but here just one value per row.


*   `True` if the row is duplicated
*   `False` if it is not duplicated

Again, we can also combine this with `.sum()` to count how many duplicated rows or `True` values exist in our database.

In [22]:
# For testing purposes, we create a small new DF

df_food = pd.DataFrame({
    'brand': ['Yum Yum', 'Yum Yum', 'Indomie', 'Indomie', 'Indomie'],
    'style': ['cup', 'cup', 'cup', 'pack', 'pack'],
    'rating': [4, 4, 3.5, 15, 5]})

In [23]:
# At first glance we can clearly see, that we have a duplicated rows
# with index numbers [0] and [1]
df_food

brand style  rating
0  Yum Yum   cup     4.0
1  Yum Yum   cup     4.0
2  Indomie   cup     3.5
3  Indomie  pack    15.0
4  Indomie  pack     5.0

In [24]:
# The method tells us, that the row with index[1] is a duplicate !
df_food.duplicated()

0    False
1     True
2    False
3    False
4    False
dtype: bool

In [25]:
# Now if we want to know the total amount of all duplicates within our DF
# We have a total of 1 duplicated row within the DF
df_food.duplicated().sum()

1

### Find out the largest/smallest values from a row within a column



*   [DataFrame.nlargest(n, columns)](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nlargest.html) returns the top `n` rows with the largest value for whatever column we specify in `columns`
*   [DataFrame.nsmallest()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nsmallest.html) returns the bottom `n` rows with the smallest value for whatever column we specify in `columns`



In [26]:
df.nlargest(5, "product_quantity")

id  id_order  product_id  product_quantity      sku unit_price  \
53860  1228150    346221           0               999  APP1190      55.99   
68712  1254032    358747           0               999  SEV0028      19.99   
57796  1234924    349475           0               800  KIN0137       7.49   
57306  1234111    349133           0               555  APP0665      70.99   
40813  1204788    335057           0               201  THU0029      80.99   

                      date  
53860  2017-04-14 21:50:52  
68712  2017-05-24 14:51:58  
57796  2017-04-25 09:59:00  
57306  2017-04-24 10:20:13  
40813  2017-03-14 15:25:53

In [27]:
df.nsmallest(5, "product_quantity")

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

## Ways to explore DataFrames:

For now we have two options to explore a DataFrame:



*   ### Attributes
  `.shape`, `.size`,`ndim` and others. Attributes are written without parentheses and give you the "raw metadata" about the DataFrame you are calling them upon



*   ### Methods
`.head()`,`.describe()`,`isna()` and others. Methods are written with parentheses and perform some sort of calculation, transformation or aggregation.
A method is like a function that is tied to a specific object type.




\



---
As always, check the [the documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html?highlight=DataFrame#pandas.DataFrame) !

---



# Select a column in a DataFrame

Given a DataFrame, we can select a particular column in several ways:



*   Indicating the name of the column between square brackets `[]`
*   With the `.loc[] attribute` (by name or tag)
*   With the `.iloc[] attribute` (by position)



If you just want to view a column, you can do that with the "plain" square brackets `[]`. Use that if you know the name of the column and don't want to modify it.

In [8]:
df["id_order"]

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

In [12]:
# This can also be paired with an index to just get a specific row
df["id_order"][2]

299541

In [13]:
# Or a slice of rows
df["id_order"][0:5]

0    299539
1    299540
2    299541
3    299542
4    299543
Name: id_order, dtype: int64

In [14]:
# You can also write
df["id_order"][:5]

0    299539
1    299540
2    299541
3    299542
4    299543
Name: id_order, dtype: int64

`.loc[]` takes two arguments `[rows, columns]`.
Passing a `:` to the rows argument means you want to "have/grab" all the rows,
basically selecting a whole column by their name.

`.loc[]` is often the prefered primary option for selecting data.

In [16]:
# Selecting a whole column by their name
df.loc[:, "id_order"]

0         299539
1         299540
2         299541
3         299542
4         299543
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293983, dtype: int64

In [18]:
# Selecting all rows up to (including) index position 5
df.loc[:5, "id_order"]

0    299539
1    299540
2    299541
3    299542
4    299543
5    295310
Name: id_order, dtype: int64

In [19]:
# Select all rows starting (including) index position 5
df.loc[5:, "id_order"]

5         295310
6         299544
7         299545
8         299546
9         295347
           ...  
293978    527398
293979    527399
293980    527400
293981    527388
293982    527401
Name: id_order, Length: 293978, dtype: int64

In [20]:
# Select only the row with an specific index position
df.loc[5, "id_order"]

295310

`.iloc[]` works similar, but only accepts integers for the two arguments that you can pass in `[rows, columns]`.

In [21]:
# Let us have a quick look again at our DF
# To make it clear, we just select the first entries.

df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [22]:
# When we now want to work with .iloc[] we basically need to count the rows/columns
# For example if we want to select the complete sku column we need to count in what position
# the sku column is, starting from left to right, beginning with index[0].
# The column "sku" would then be index position[4]

df.iloc[:, 4]

0         OTT0133
1         LGE0043
2         PAR0071
3         WDT0315
4         JBL0104
           ...   
293978    JBL0122
293979    PAC0653
293980    APP0698
293981    BEZ0204
293982    APP0927
Name: sku, Length: 293983, dtype: object

# Selecting multiple columns

Selecting multiple columns works with all of the above listed options.
The difference is, that we need to make some slight modifications in some regards.

Multiple columns are by definition a list. 
A list in Python is represented with `[]`, so we need for our `.loc[]` and `.iloc[]` 2 sets of square brackets.

In [34]:
# Select two columns "id_order" & "sku"
df[["id_order","sku"]]

id_order      sku
0         299539  OTT0133
1         299540  LGE0043
2         299541  PAR0071
3         299542  WDT0315
4         299543  JBL0104
...          ...      ...
293978    527398  JBL0122
293979    527399  PAC0653
293980    527400  APP0698
293981    527388  BEZ0204
293982    527401  APP0927

[293983 rows x 2 columns]